File: solutions/zig/perfect-numbers/1/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};
const std = @import("std");
/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    std.debug.assert(n != 0);
    var s: u64 = 0;
    var i: u64 = 1;
    while (i * i < n) : (i += 1) {
        if (n % i == 0) {
            s += i + n / i;
        }
    }
    if (i * i == n) {
        s += i;
    }
    s -= n;
    if (s == n) {
        return Classification.perfect;
    } else if (s < n) {
        return Classification.deficient;
    } else {
        return Classification.abundant;
    }
}

File: solutions/ocaml/perfect-numbers/1/perfect_numbers.ml
let classify n = 
  if n <= 0 then Error "Classification is only possible for positive integers." else 
    let s = List.fold_left (fun acc a -> acc + a) 0 (List.filter (fun x -> n mod x = 0) (List.init (n-1) (fun i -> i+1))) in
    if s = n then Ok "perfect" else
      if s < n then Ok "deficient" else Ok "abundant"